content/posts/pandastutor.md

```diff
@@ -0,0 +1,313 @@
+
+The recommended way to load custom Python code into Pyodide is [by using
+a
+wheel](https://pyodide.org/en/stable/usage/loading-custom-python-code.html),
+so we first created a wheel to package up the Pandas Tutor backend
+code. That was straightforward since it's pure-Python and doesn't
+require any WASM code compilation.
+
+However, we ran into our first issue right away: Since Pandas Tutor has
+almost a dozen package dependencies, we originally thought we could use
+`micropip` to install them all, as the [Pyodide documentation
+recommends](https://pyodide.org/en/stable/usage/loading-packages.html).
+Unfortunately, some packages couldn't be installed with `micropip` since
+they didn't have wheels available on PyPI (e.g.,
+`micropip.install('docopt')` doesn't work). Our solution was to bundle
+(i.e., inline) all dependencies into a self-contained Pandas
+Tutor wheel so that our web frontend code needs only one call to
+`micropip` instead of almost a dozen separate calls:
+
+```js
+const micropip = pyodide.pyimport("micropip");
+await micropip.install("https://pandastutor.com/build/pandastutor-1.0-py3-none-any.whl");
+```
+
+This approach also has the nice side-effect of pinning all the
+dependency versions for better API stability. The only dependency we
+didn't bundle was pandas because that requires WASM-compiled code. But
+fortunately Pyodide comes with pandas
+[built-in](https://pyodide.org/en/stable/usage/packages-in-pyodide.html)!
+
+
+### Step 2: import the Pandas Tutor module in JavaScript
+
+We added one line to our frontend JavaScript to import `main.py` from
+Pandas Tutor when the page loads (after the wheel gets installed from
+Step 1):
+
+```js
+const pandastutor_py = pyodide.pyimport("pandas_tutor.main");
+```
+
+This is a synchronous call that imports `pandas_tutor` and all
+dependencies when the web app first loads, so it takes 3-5 seconds
+to finish since pandas and other libraries must be loaded.
+Unfortunately, browser caching can't speed things up here since even if
+all files were cached locally, Pyodide still needs to run the import
+and initialization code.
+
+One feature addition that might speed this up is if Pyodide
+could *serialize a binary memory snapshot* taken right after
+everything gets initialized. Then the user's browser can directly load
+that snapshot on the first page visit and cache it for future visits.
+(But maybe that snapshot would be too big to load efficiently.
+[Related GitHub
+Issues](https://github.com/pyodide/pyodide/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+size+label%3A%22size+%26+load+time+optimization%22).)
+
+
+### Step 3: call Pandas Tutor backend code from JavaScript
+
+Now we can call the Pandas Tutor backend from our frontend JavaScript
+with just one line:
+
+```js
+// 1) frontend grabs userCode from the user's in-browser code editor
+// 2) run the user's code to produce an execution trace ...
+      const executionTrace = pandastutor_py.run_user_code(userCode);  
```

Comment:
```suggestion
const executionTrace = pandastutor_py.run_user_code(userCode);  
```
The indentation with respect to the comments looks a bit strange otherwise.